<class name="error ctx" context="public"  lifecycle="none">
    Error context.
    Can be used for sequential operations, i.e. parsers, to accumulate error.
    In this way operation is successful if all steps are successful, otherwise
    last occurred error code can be obtained.

    <property name="error" type="error"/>


    <method name="reset">
        Reset context to the "no error" state.

        <argument class="context"/>
    </method>


    <method name="update">
        Update context with given error.

        <argument class="context"/>
        <argument name="error" type="error"/>
    </method>


    <macros name="safe update">
        Perform update only if context defined, otherwise log error.

        <code>
            #define .(c_class_error_ctx_macros_safe_update) (CTX, ERR)
                do {
                    if (NULL != (CTX)) {
                        .(c_class_error_ctx_method_update) ((CTX), (ERR));
                    } else {
                        /* TODO: Log this error, when logging will be added. */
                    }
                } while (false)
        </code>
    </macros>


    <method name="error">
        Reset context to the "no error" state.

        <argument class="context" access="readonly"/>
        <return type="error"/>
    </method>
</class>
